File: windows-app/Plantoir.Core/Models/CloudSyncedFolder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace Plantoir.Core.Models;

/// <summary>
/// Whether a working folder is kept in sync by a cloud service, and which one.
///
/// <para><b>A synced folder is never refused.</b> Teachers keep their notes in
/// one on purpose — it is how the notes reach an iPad and a second machine — so
/// Plantoir recognises it, says once and in plain words what it costs, lets the
/// teacher decide, and leaves their content where they put it. Refusing was
/// considered and rejected: a hard block is the one answer a teacher cannot opt
/// out of, and it would mean telling them to give up reaching their own notes
/// from other devices. Detection is unreliable in both directions besides — a
/// false refusal is unrecoverable for them, and a synced folder Plantoir cannot
/// see still works.</para>
///
/// <para><b>From markers the system exposes, never from a folder's NAME.</b> A
/// teacher can have a folder literally called "Dropbox" that is not one, and a
/// wrong answer here is shown to them as fact.</para>
///
/// <para>The Windows markers differ from the mac's and are deliberately not
/// shared — see <c>shared-rules.json</c> → <c>cloudSyncedFolders.detection</c>,
/// whose eleven cases are all mac paths. What IS shared is everything a teacher
/// reads and every rule about when they read it.</para>
/// </summary>
public static class CloudSyncedFolder
{
    /// <summary>
    /// What a service is called when the folder is recognisably synced but the
    /// service is not one Plantoir names. Saying so honestly beats guessing.
    /// </summary>
    public const string UnknownServiceName = "your cloud service";

    /// <summary>
    /// The service syncing this folder, or null when nothing says one is.
    ///
    /// <para>Null is a perfectly ordinary answer and never blocks anything: a
    /// service whose client publishes no marker Plantoir can read is simply
    /// unrecognised, which the contract explicitly allows.</para>
    /// </summary>
    public static string? ServiceFor(string? path) => ServiceFor(path, new EnvironmentReader());

    /// <summary>
    /// The rule, with the environment injected so it can be tested without
    /// installing five cloud clients.
    /// </summary>
    public static string? ServiceFor(string? path, IEnvironmentReader environment)
    {
        if (string.IsNullOrWhiteSpace(path)) return null;

        string full;
        try { full = Path.GetFullPath(path).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar); }
        catch { return null; }
        if (full.Length == 0) return null;

        // OneDrive publishes its roots in the environment. A Desktop or
        // Documents folder moved into OneDrive by "Known Folder Move"
        // physically lives under one of these, so the prefix rule catches it
        // without a special case.
        foreach (string variable in new[] { "OneDrive", "OneDriveConsumer", "OneDriveCommercial" })
        {
            string? root = environment.Variable(variable);
            if (IsInside(full, root)) return "OneDrive";
        }

        // Dropbox records every synced root, personal and business, in its own
        // info.json. Read rather than guessed: a teacher can move the Dropbox
        // folder anywhere.
        foreach (string root in DropboxRoots(environment))
            if (IsInside(full, root)) return "Dropbox";

        // iCloud for Windows, at its default location. The client's own
        // setting would be better and is not reliably readable; an unrecognised
        // custom location is allowed.
        string? profile = environment.Variable("USERPROFILE");
        if (!string.IsNullOrEmpty(profile) && IsInside(full, Path.Combine(profile!, "iCloudDrive")))
            return "iCloud Drive";

        // Google Drive and Box are recognised only where their clients publish
        // a mount point in the environment. Otherwise unrecognised — allowed.
        if (IsInside(full, environment.Variable("GoogleDriveMount"))) return "Google Drive";
        if (IsInside(full, environment.Variable("BoxDrive"))) return "Box";

        return null;
    }

    /// <summary>Whether this folder is synced by anything Plantoir can see.</summary>
    public static bool IsSynced(string? path) => ServiceFor(path) is not null;

    /// <summary>
    /// Every root Dropbox says it syncs, from its own info.json.
    ///
    /// <para>Both locations are read because Dropbox has used each: the file is
    /// in <c>%APPDATA%</c> for some installs and <c>%LOCALAPPDATA%</c> for
    /// others. A malformed or unreadable file yields nothing rather than
    /// throwing — this runs while a teacher is choosing a folder, and a crash
    /// there would be far worse than an unrecognised service.</para>
    /// </summary>
    private static IEnumerable<string> DropboxRoots(IEnvironmentReader environment)
    {
        var roots = new List<string>();
        foreach (string variable in new[] { "APPDATA", "LOCALAPPDATA" })
        {
            string? baseDir = environment.Variable(variable);
            if (string.IsNullOrEmpty(baseDir)) continue;
            string info = Path.Combine(baseDir!, "Dropbox", "info.json");
            string text;
            try
            {
                if (!environment.FileExists(info)) continue;
                text = environment.ReadAllText(info);
            }
            catch { continue; }

            try
            {
                using var document = JsonDocument.Parse(text);
                if (document.RootElement.ValueKind != JsonValueKind.Object) continue;
                // { "personal": { "path": "..." }, "business": { "path": "..." } }
                foreach (var account in document.RootElement.EnumerateObject())
                {
                    if (account.Value.ValueKind != JsonValueKind.Object) continue;
                    if (account.Value.TryGetProperty("path", out var p) && p.ValueKind == JsonValueKind.String)
                    {
                        string? root = p.GetString();
                        if (!string.IsNullOrWhiteSpace(root)) roots.Add(root!);
                    }
                }
            }
            catch (JsonException) { continue; }
        }
        return roots;
    }

    /// <summary>
    /// Whether <paramref name="candidate"/> IS <paramref name="root"/> or sits
    /// inside it — ending at a path boundary, never as a bare substring, so
    /// <c>C:\Users\x\OneDrive2</c> is not inside <c>C:\Users\x\OneDrive</c>.
    /// Case-insensitive, because Windows filesystems are.
    /// </summary>
    private static bool IsInside(string candidate, string? root)
    {
        if (string.IsNullOrWhiteSpace(root)) return false;
        string normalised;
        try { normalised = Path.GetFullPath(root!).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar); }
        catch { return false; }
        if (normalised.Length == 0) return false;

        if (candidate.Equals(normalised, StringComparison.OrdinalIgnoreCase)) return true;
        return candidate.StartsWith(normalised + Path.DirectorySeparatorChar, StringComparison.OrdinalIgnoreCase)
            || candidate.StartsWith(normalised + Path.AltDirectorySeparatorChar, StringComparison.OrdinalIgnoreCase);
    }
}

/// <summary>
/// The bits of the machine the rule asks about. Injected so the detection can
/// be tested without five cloud clients installed.
/// </summary>
public interface IEnvironmentReader
{
    string? Variable(string name);
    bool FileExists(string path);
    string ReadAllText(string path);
}

/// <summary>The real machine.</summary>
public sealed class EnvironmentReader : IEnvironmentReader
{
    public string? Variable(string name) => Environment.GetEnvironmentVariable(name);
    public bool FileExists(string path) => File.Exists(path);
    public string ReadAllText(string path) => File.ReadAllText(path);
}

File: mac-app/QuartzTeachers/Models/CloudSyncedFolder.swift
import Foundation

/// A working folder that a cloud service keeps in sync — iCloud Drive,
/// Dropbox, OneDrive, Google Drive and the like.
///
/// Teachers keep their notes in such folders ON PURPOSE: it is how the notes
/// reach an iPad and a second Mac. So Plantoir never refuses one. What it does
/// is SAY so, once, in plain words — because three things go wrong in a synced
/// folder that a teacher would otherwise meet as unexplained slowness:
///
/// 1. A build writes thousands of small files, and the sync app copies every
///    one of them to the cloud.
/// 2. A page the service has offloaded to save space has to be downloaded
///    before it can be read, and renaming a folder reads every page.
/// 3. A move or rename can fail while the service is busy copying that file.
///
/// The decision, made 2026-09-05 (see `TODO.md` for both sides of it): detect,
/// explain, let the teacher choose, and leave their content where they put it.
/// Refusing was rejected because a hard block is the one answer a teacher
/// cannot opt out of, and detection is not certain enough to block on.
nonisolated struct CloudSyncedFolder: Equatable, Sendable {

    // MARK: - Stored properties

    /// The service, in the words a teacher uses for it: "iCloud Drive",
    /// "Dropbox", "OneDrive", "Google Drive", "Box" — or, when the folder is
    /// recognisably synced but the service is not one Plantoir knows by name,
    /// "your cloud service".
    let serviceName: String

    /// The folder itself.
    let folderPath: String
}

/// Recognises a synced folder from markers the system exposes, never from a
/// guess at a name — a teacher can have a folder literally called "Dropbox"
/// that is not one, and a wrong answer here is shown to them as fact.
///
/// The pure rule (`serviceName(forPath:homePath:)`) is what the contract
/// pins; `syncedFolder(at:)` adds the one check that needs the live file
/// system — iCloud's own flag on the item — which catches a Desktop or
/// Documents folder that iCloud syncs in place.
nonisolated enum CloudSyncDetector {

    // MARK: - Stored properties

    /// The name used when the folder is synced by a service not listed below.
    static let unknownServiceName: String = "your cloud service"

    /// Where macOS keeps every File Provider-based service's folders since
    /// macOS 12.3: `~/Library/CloudStorage/<Service>-<Account>/`. Dropbox,
    /// OneDrive, Google Drive and Box all live here now, which is what makes
    /// this detectable at all.
    static let cloudStorageFolderName: String = "Library/CloudStorage"

    /// iCloud Drive's real location; `~/iCloud Drive` in Finder is a view of it.
    static let iCloudFolderName: String = "Library/Mobile Documents"

    /// How a `~/Library/CloudStorage/` entry's prefix maps to the name a
    /// teacher knows. The prefix is the part before the first hyphen:
    /// `OneDrive-Personal`, `GoogleDrive-name@gmail.com`, `Dropbox`,
    /// `Box-Box`.
    static let serviceNamesByFolderPrefix: [String: String] = [
        "OneDrive": "OneDrive",
        "GoogleDrive": "Google Drive",
        "Dropbox": "Dropbox",
        "Box": "Box",
    ]

    // MARK: - Functions

    /// The service syncing this folder, or nil when no marker says one does.
    ///
    /// Symlinks are resolved FIRST. Dropbox and OneDrive both leave a link
    /// at the old place (`~/Dropbox` → `~/Library/CloudStorage/Dropbox`), and
    /// a path arriving through it matches no rule while the folder behind it
    /// matches Dropbox's. Resolving also makes one folder one key for the
    /// acknowledgement, whichever spelling it arrived by.
    static func syncedFolder(at folderURL: URL, homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser) -> CloudSyncedFolder? {
        let resolvedURL: URL = folderURL.standardizedFileURL.resolvingSymlinksInPath()
        let folderPath: String = resolvedURL.path
        let homePath: String = homeDirectory.standardizedFileURL.resolvingSymlinksInPath().path
        if let serviceName = CloudSyncDetector.serviceName(forPath: folderPath, homePath: homePath) {
            return CloudSyncedFolder(serviceName: serviceName, folderPath: folderPath)
        }
        if CloudSyncDetector.iCloudFlagIsTrusted(forPath: folderPath, homePath: homePath) && CloudSyncDetector.isInICloud(resolvedURL) {
            return CloudSyncedFolder(serviceName: "iCloud Drive", folderPath: folderPath)
        }
        return nil
    }

    /// Where the item's own iCloud flag is believed: under `~/Desktop` and
    /// `~/Documents`, the two folders "Desktop & Documents Folders" syncs in
    /// place without moving them. Nowhere else — the flag is set for ANY
    /// File Provider item, a Dropbox or OneDrive folder included (checked on
    /// a real `~/Library/CloudStorage/Dropbox`), so trusting it everywhere
    /// would call a Dropbox folder "iCloud Drive" to the teacher's face.
    static func iCloudFlagIsTrusted(forPath path: String, homePath: String) -> Bool {
        let home: String = CloudSyncDetector.withoutTrailingSlash(homePath)
        for folderName in ["Desktop", "Documents"] {
            let root: String = home + "/" + folderName
            if path == root || path.hasPrefix(root + "/") {
                return true
            }
        }
        return false
    }

    /// The pure rule: which service, from the path alone.
    ///
    /// Answers for the two locations macOS reserves for synced folders, and
    /// nothing else. A folder called "Dropbox" on the Desktop is NOT matched:
    /// the name proves nothing, and the File Provider location does.
    static func serviceName(forPath path: String, homePath: String) -> String? {
        let home: String = CloudSyncDetector.withoutTrailingSlash(homePath)
        let iCloudRoot: String = home + "/" + CloudSyncDetector.iCloudFolderName
        if path == iCloudRoot || path.hasPrefix(iCloudRoot + "/") {
            return "iCloud Drive"
        }

        let cloudStorageRoot: String = home + "/" + CloudSyncDetector.cloudStorageFolderName + "/"
        if !path.hasPrefix(cloudStorageRoot) {
            return nil
        }
        // `<Service>-<Account>/...` — the first component after the root.
        let remainder: String = String(path.dropFirst(cloudStorageRoot.count))
        var serviceFolderName: String = ""
        for character in remainder {
            if character == "/" {
                break
            }
            serviceFolderName.append(character)
        }
        if serviceFolderName.isEmpty {
            return nil
        }
        var prefix: String = ""
        for character in serviceFolderName {
            if character == "-" {
                break
            }
            prefix.append(character)
        }
        if let knownName = CloudSyncDetector.serviceNamesByFolderPrefix[prefix] {
            return knownName
        }
        return CloudSyncDetector.unknownServiceName
    }

    /// The item's own "ubiquitous" flag — the only way to see a Desktop or
    /// Documents folder that "Desktop & Documents Folders" syncs in place,
    /// since those stay at `~/Desktop` and `~/Documents` and match no path
    /// rule. Only meaningful where `iCloudFlagIsTrusted` says so.
    static func isInICloud(_ folderURL: URL) -> Bool {
        guard let values = try? folderURL.resourceValues(forKeys: [.isUbiquitousItemKey]) else {
            return false
        }
        return values.isUbiquitousItem ?? false
    }

    static func withoutTrailingSlash(_ path: String) -> String {
        if path.hasSuffix("/") && path.count > 1 {
            return String(path.dropLast())
        }
        return path
    }
}

/// What a teacher is told about a synced working folder — written once, here,
/// and pinned by `contracts/shared-rules.json` → `cloudSyncedFolders.wording`
/// so Windows says the same things.
///
/// The words are chosen to name EFFECTS a teacher can recognise ("building can
/// be slower", "renaming a folder can take a while") and never machinery. No
/// "sync client", no "file provider", no "dataless file".
nonisolated enum CloudSyncWording {

    // MARK: - Functions

    /// The headline, said first in both the picker and the notice.
    static func headline(service: String) -> String {
        return "This folder is kept in sync with \(service)."
    }

    /// The one-line version for the notice a window shows.
    static let summary: String =
        "Plantoir works here, but building can be slower and renaming folders can take a while. Your notes stay where they are."

    /// The reassurance, said before anything alarming.
    static let notesStayPut: String =
        "Your course notes stay exactly where they are. Plantoir never moves them."

    /// Offloaded pages: slow, not harmful.
    static func offloadedPagesAreSlow(service: String) -> String {
        return "If \(service) has moved some of your pages off this computer to save space, Plantoir has to download each one before it can read it. Renaming a folder reads every page, so it can take a while."
    }

    /// A move or rename can fail mid-sync; nothing is lost, and it can be
    /// tried again.
    static func syncingCanInterruptAMove(service: String) -> String {
        return "If \(service) is busy copying a file at the moment Plantoir needs to move or rename it, that step can fail. Plantoir checks before it moves anything, and you can try again once syncing has settled."
    }

    /// The choice, stated without pushing either way: a folder that is not
    /// synced avoids all of it; a folder that reaches other devices may
    /// matter more.
    static let whatToDo: String =
        "To avoid all of this, keep your working folder somewhere that is not synced — a folder inside your home folder, for example. If reaching your notes from other devices matters more, carry on: Plantoir will work, just more slowly at times."

    /// The button that goes ahead in the picker.
    static let useAnywayButton: String = "Use This Folder Anyway"

    /// The button that dismisses the in-window notice.
    static let dismissNoticeButton: String = "Got It"

    /// The picker's other button — the same words it has always used, named
    /// here so the contract carries them beside the choice they complete.
    static let chooseDifferentFolderButton: String = "Choose a Different Folder…"

    /// The notice's button that opens the full explanation in place, and
    /// what it says once the explanation is open.
    static let showDetailsButton: String = "Show Details"
    static let hideDetailsButton: String = "Hide Details"

    /// The full explanation, in the order it is shown.
    static func explanation(service: String) -> [String] {
        return [
            CloudSyncWording.notesStayPut,
            // The sentence about thousands of build files being copied to the
            // cloud used to sit here, and it left on 2026-09-05 when the built
            // site moved out of the working folder (BuildOutputLocation). It
            // is not "one fewer warning": it was the only one of the four that
            // was no longer TRUE, and a warning a teacher can check and find
            // false costs the other three their credit. See
            // contracts/shared-rules.json -> cloudSyncedFolders.wording
            // .buildFilesAreCopiedRetired.
            CloudSyncWording.offloadedPagesAreSlow(service: service),
            CloudSyncWording.syncingCanInterruptAMove(service: service),
            CloudSyncWording.whatToDo,
        ]
    }
}
